tm2/pkg/sdk/params/keeper.go

```diff
@@ -145,34 +153,85 @@ func (pk ParamsKeeper) SetBytes(ctx sdk.Context, key string, value []byte) {
 	pk.set(ctx, key, value)
 }
 
-func (pk ParamsKeeper) getIfExists(ctx sdk.Context, key string, ptr interface{}) {
+// GetParam gets a param value from the global param store.
+// Users generally should not cache anything and instead rely on the efficiency
+// of paramk.GetParams().
+func (pk ParamsKeeper) GetParams(ctx sdk.Context, moduleKey string, key string, target interface{}) (bool, error) {
+	if moduleKey != "" {
+		if pk.IsRegistered(moduleKey) {
+			key = moduleKey + ":" + key
+		} else {
+			return false, fmt.Errorf("params module key %q does not exisit", moduleKey)
+		}
+	}
+
 	stor := ctx.Store(pk.key)
-	bz := stor.Get([]byte(key))
+	vk := ValueStoreKey(key)
+	bz := stor.Get(vk)
 	if bz == nil {
-		return
+		return false, nil
 	}
-	err := amino.UnmarshalJSON(bz, ptr)
+
+	return true, amino.UnmarshalJSON(bz, target)
+}
+
+// SetParam sets a param value to the global param store.
+func (pk ParamsKeeper) SetParams(ctx sdk.Context, moduleKey string, key string, param interface{}) error {
+	if moduleKey != "" {
+		if pk.IsRegistered(moduleKey) {
+			key = moduleKey + ":" + key
+		} else {
+			return fmt.Errorf("parameter module key %q does not exist", moduleKey)
+		}
+	}
+
+	bz, err := amino.MarshalJSON(param)
 	if err != nil {
-		panic(err)
+		return err
 	}
-}
 
-func (pk ParamsKeeper) get(ctx sdk.Context, key string, ptr interface{}) {
 	stor := ctx.Store(pk.key)
-	bz := stor.Get([]byte(key))
-	err := amino.UnmarshalJSON(bz, ptr)
+	vk := ValueStoreKey(key)
+	stor.Set(vk, bz)
+	return nil
+}
+
+func (pk ParamsKeeper) getIfExists(ctx sdk.Context, key string, ptr interface{}) {
+	module, rawKey := parsePrefix(key)
+	_, err := pk.GetParams(ctx, module, rawKey, ptr)
```

Comment:
wtf

Reply:
The key passed in here could contain a prefixed module  name as part of the key, which called from GetXXX(). I parse it out and validate it in GetParam, rather than simply returning nothing when the provided prefixed module does not exist, as is done in the current implementation on the master branch. Additionally, pk.GetParams handles both individual attribute parameters and parameters as a struct.

Reply:
I know, but do you see the problem?